### PR TITLE
[Merged by Bors] - reflect: add `insert` and `remove` methods to `List`

### LINKED ISSUE #7061: reflect: `insert` and `remove` methods for `List`
## What problem does this solve or what need does it fill?

Currently, the reflect `List` trait implements only `push` and `pop` functionality. There is no way to insert or remove elements other than from/to the end other than by downcasting. This is especially notable using types like `VecDeque` with reflection.

## What solution would you like?

Add `insert` and `remove` methods for `List` and implement them.
Give `push` and `pop` default implementations using on the `insert`/`remove` and `len` methods.

## What alternative(s) have you considered?

We might not need to implement the methods by default but since we can I believe we should as to not put to much burden on the implementor.

## Additional context

Discussed first [on discord](https://discord.com/channels/691052431525675048/1002362493634629796/1054711227454078977).

### PR DESCRIPTION
**Pull request by soqb**
# Objective

- Fixes #7061

## Solution

- Add and implement `insert` and `remove` methods for `List`.

---

## Changelog

- Added `insert` and `remove` methods to `List`.
- Changed the `push` and `pop` methods on `List` to have default implementations.

## Migration Guide

- Manual implementors of `List` need to implement the new methods `insert` and `remove` and 
consider whether to use the new default implementation of `push` and `pop`.

### CODE REVIEW

**Review by alice-i-cecile — APPROVED** [review]

File: crates/bevy_reflect/src/impls/std.rs [hunk 1]
```diff
@@ -213,6 +213,22 @@ macro_rules! impl_reflect_for_veclike {
...
+                            "Attempted to insert invalid value of type {}.",
+                            value.type_name()
+                        )
+                    })
+                });
+                $insert(self, index, value);
+            }
+
+            fn remove(&mut self, index: usize) -> Box<dyn Reflect> {
+                Box::new($remove(self, index)) as Box<dyn Reflect>
```
Comment (MrGVSV) [inline crates/bevy_reflect/src/impls/std.rs:229]: Nit: I know the other methods already have it, but I don't think this cast is necessary, right?
```suggestion
                Box::new($remove(self, index))
```

File: crates/bevy_reflect/src/impls/smallvec.rs [hunk 2]
```diff
@@ -49,6 +49,22 @@ impl<T: smallvec::Array + Send + Sync + 'static> List for SmallVec<T>
...
+                    "Attempted to insert invalid value of type {}.",
+                    value.type_name()
+                )
+            })
+        });
+        SmallVec::insert(self, index, value);
+    }
+
+    fn remove(&mut self, index: usize) -> Box<dyn Reflect> {
+        Box::new(self.remove(index)) as Box<dyn Reflect>
```
Comment (MrGVSV) [inline crates/bevy_reflect/src/impls/smallvec.rs:65]: Same here
```suggestion
        Box::new(self.remove(index))
```

File: crates/bevy_reflect/src/list.rs [hunk 3]
```diff
@@ -9,18 +9,41 @@ use crate::{
...
+/// allowing it's internal size to grow.
 ///
 /// This trait expects index 0 to contain the _front_ element.
 /// The _back_ element must refer to the element with the largest index.
 /// These two rules above should be upheld by manual implementors.
+///
+/// [`push`](List::push) and [`pop`](List::pop) have default implementations,
+/// however it may be faster to implement them manually.
 pub trait List: Reflect + Array {
+    /// Inserts an element at position `index` within the list.
```
Comment (MrGVSV) [inline crates/bevy_reflect/src/list.rs:22]: Nit: Should we include the rest of `Vec`'s [doc](https://doc.rust-lang.org/std/vec/struct.Vec.html#method.insert) here? I know the elements might actually be shifted in memory, but their position within the list will reflect the shift.
```suggestion
    /// Inserts an element at position `index` within the list, shifting all elements after it to the right.
```

File: crates/bevy_reflect/src/list.rs [hunk 4]
```diff
@@ -9,18 +9,41 @@ use crate::{
...
+    ///
+    /// # Panics
+    /// Panics if `index > len`.
+    fn insert(&mut self, index: usize, element: Box<dyn Reflect>);
+
+    /// Removes and returns the element at position `index` within the list.
```
Comment (MrGVSV) [inline crates/bevy_reflect/src/list.rs:28]: Nit: Similarly, should we update these docs to [match](https://doc.rust-lang.org/std/vec/struct.Vec.html#method.remove) as well?
```suggestion
    /// Removes and returns the element at position `index` within the list, shifting all elements after it to the left.
```

**Review by MrGVSV — APPROVED** [review]
LGTM! Just a few non-blocking comments

File: crates/bevy_reflect/src/list.rs (hunk 3 above)
Comment (soqb) [reply crates/bevy_reflect/src/list.rs:22]: i don't think the `List` trait has a concept of "to the right". `Vec`'s docs are explaining the change in memory layout where the bytes containing elements after the index are physically shifted to be to the right of the new element, whereas here we make no guarantee that's what is done to the memory.

perhaps we should have something more inline with [`VecDeque`'s docs](https://doc.rust-lang.org/stable/std/collections/struct.VecDeque.html#method.insert).

```suggestion
    // Inserts an element at position `index` within the list,
    // shifting all elements after it towards the back of the list.
```

File: crates/bevy_reflect/src/list.rs (hunk 4 above)
Comment (soqb) [reply crates/bevy_reflect/src/list.rs:29]:
```suggestion
    /// Removes and returns an element at position `index` within the list,
    /// shifting all elements after it towards the front of the list.
```

**Comment by soqb** [comment]
surely this should have `C-Breaking-Breaking, no?

File: crates/bevy_reflect/src/list.rs (hunk 3 above)
Comment (MrGVSV) [reply crates/bevy_reflect/src/list.rs]: Good point. `VecDeque`'s docs are definitely more inline with the terminology we've been using for this trait.

**Review by JoJoJet — APPROVED** [review]

File: crates/bevy_reflect/src/list.rs [hunk 5]
```diff
@@ -9,18 +9,43 @@ use crate::{
 
 /// An ordered, mutable list of [Reflect] items. This corresponds to types like [`std::vec::Vec`].
 ///
-/// This is a sub-trait of [`Array`] as it implements a [`push`](List::push) function, allowing
-/// it's internal size to grow.
+/// This is a sub-trait of [`Array`] as it implements [insertion](List::insert) and [removal](List::remove),
+/// allowing it's internal size to grow.
```
Comment (johanhelsing) [inline crates/bevy_reflect/src/list.rs:13]:
```suggestion
/// This is a sub-trait of [`Array`], however as it implements [insertion](List::insert) and [removal](List::remove),
/// it's internal size may change.
```

**Review by johanhelsing — COMMENTED** [review]
Looks good to me :) One small nit for the docs.

After this gets in, should we consider making `apply` remove elements if `self.len() > value.len()`? Ref: #7129

**Review by johanhelsing — APPROVED** [review]

**Review by alice-i-cecile — COMMENTED** [review]
bors r+

**Pull request closed without merging**